File: test-verification/compare_data_values.py
"""
Deep Data Comparison - Compare actual REF values (Python)
Verifies that decompression and scaling produce identical numerical results
"""

import pyart
from pyart.io.nexrad_level2 import NEXRADLevel2File
from pyart.io.common import prepare_for_read
import json
import numpy as np

# Test with KHGX file (most comprehensive)
TEST_FILE = 's3://unidata-nexrad-level2/2022/03/22/KHGX/KHGX20220322_120125_V06'

def compare_data_values():
    print('=' * 80)
    print('Deep Data Value Comparison: Python/PyART')
    print('=' * 80)
    print(f'\nTest File: {TEST_FILE}\n')

    # Read with Python
    print('Reading with Python/PyART...')
    file_handle = prepare_for_read(TEST_FILE, storage_options={'anon': True})
    nexrad_file = NEXRADLevel2File(file_handle)

    # Get REF data
    print('Extracting REF moment data...')
    range_info = nexrad_file.get_range(0, 'REF')
    max_ngates = len(range_info)
    ref_data = nexrad_file.get_data('REF', max_ngates)

    print(f'\nPython REF Data Shape: {ref_data.shape}')

    # Calculate statistics for all radials
    stats = {
        'total_radials': ref_data.shape[0],
        'total_gates': ref_data.shape[1],
        'total_values': ref_data.size,
        'valid_values': 0,
        'null_values': 0,

        # Per-radial statistics
        'radial_samples': [],

        # Overall statistics (excluding nulls/masked)
        'all_valid_values': [],
        'min': None,
        'max': None,
        'mean': None,

        # Sample radials for detailed comparison
        'sample_radials': [0, 100, 500, 1000, 5000, 10000]
    }

    # Collect statistics
    for radial in range(ref_data.shape[0]):
        radial_data = ref_data[radial, :]

        # Count valid vs masked/null values
        if np.ma.is_masked(radial_data):
            valid_mask = ~radial_data.mask
            valid_values = radial_data[valid_mask]
            valid_count = valid_values.size
            null_count = radial_data.size - valid_count
        else:
            valid_mask = ~np.isnan(radial_data)
            valid_values = radial_data[valid_mask]
            valid_count = valid_values.size
            null_count = radial_data.size - valid_count

        stats['valid_values'] += valid_count
        stats['null_values'] += null_count

        # Store sample radials
        if radial in stats['sample_radials']:
            radial_min = float(np.min(valid_values)) if valid_count > 0 else None
            radial_max = float(np.max(valid_values)) if valid_count > 0 else None
            radial_mean = float(np.mean(valid_values)) if valid_count > 0 else None

            # Convert first 20 values to list (handle masked values)
            values_sample = []
            for i in range(min(20, radial_data.size)):
                val = radial_data[i]
                if np.ma.is_masked(val) or np.isnan(val):
                    values_sample.append(None)
                else:
                    values_sample.append(float(val))

            # Get specific gate values
            gate_samples = {}
            for gate_idx in [0, 100, 500, 1000, 1500]:
                if gate_idx < radial_data.size:
                    val = radial_data[gate_idx]
                    if np.ma.is_masked(val) or np.isnan(val):
                        gate_samples[f'gate_{gate_idx}'] = None
                    else:
                        gate_samples[f'gate_{gate_idx}'] = float(val)

            stats['radial_samples'].append({
                'radial_index': radial,
                'total_gates': int(radial_data.size),
                'valid_count': int(valid_count),
                'null_count': int(null_count),
                'min': radial_min,
                'max': radial_max,
                'mean': radial_mean,
                'values_sample': values_sample,
                'gate_samples': gate_samples
            })

        # Add valid values to overall list
        if valid_count > 0:
            stats['all_valid_values'].extend(valid_values.tolist())

    # Calculate overall statistics
    if len(stats['all_valid_values']) > 0:
        stats['min'] = float(np.min(stats['all_valid_values']))
        stats['max'] = float(np.max(stats['all_valid_values']))
        stats['mean'] = float(np.mean(stats['all_valid_values']))

    # Print summary statistics
    print('\n' + '=' * 80)
    print('Python REF Data Statistics')
    print('=' * 80)
    print(f"Total Values: {stats['total_values']:,}")
    print(f"Valid Values: {stats['valid_values']:,} ({stats['valid_values'] / stats['total_values'] * 100:.2f}%)")
    print(f"Null Values: {stats['null_values']:,} ({stats['null_values'] / stats['total_values'] * 100:.2f}%)")
    print(f"\nOverall Statistics (valid values only):")
    print(f"  Min: {stats['min']:.4f}" if stats['min'] is not None else "  Min: None")
    print(f"  Max: {stats['max']:.4f}" if stats['max'] is not None else "  Max: None")
    print(f"  Mean: {stats['mean']:.4f}" if stats['mean'] is not None else "  Mean: None")

    # Print sample radials
    print('\n' + '=' * 80)
    print('Sample Radials (for JavaScript comparison)')
    print('=' * 80)

    for sample in stats['radial_samples']:
        print(f"\nRadial {sample['radial_index']}:")
        print(f"  Valid: {sample['valid_count']}, Null: {sample['null_count']}")
        if sample['min'] is not None:
            print(f"  Min: {sample['min']:.4f}, Max: {sample['max']:.4f}, Mean: {sample['mean']:.4f}")
        values_str = ', '.join([
            'null' if v is None else f"{v:.2f}"
            for v in sample['values_sample'][:10]
        ])
        print(f"  First 10 values: [{values_str}]")
        print(f"  Gate samples:")
        for gate, value in sample['gate_samples'].items():
            val_str = 'null' if value is None else f"{value:.4f}"
            print(f"    {gate}: {val_str}")

    # Save detailed results for comparison
    # Convert first radial to list
    first_radial_list = []
    for i in range(ref_data.shape[1]):
        val = ref_data[0, i]
        if np.ma.is_masked(val) or np.isnan(val):
            first_radial_list.append(None)
        else:
            first_radial_list.append(float(val))

    output = {
        'test_file': TEST_FILE,
        'data_shape': list(ref_data.shape),
        'statistics': {
            'total_values': stats['total_values'],
            'valid_values': stats['valid_values'],
            'null_values': stats['null_values'],
            'min': stats['min'],
            'max': stats['max'],
            'mean': stats['mean']
        },
        'radial_samples': stats['radial_samples'],
        'first_radial_full': first_radial_list
    }

    output_path = 'output/data_comparison_py.json'
    with open(output_path, 'w') as f:
        json.dump(output, f, indent=2)

    print(f'\n[OK] Detailed results saved to: {output_path}')

    # Compare with JavaScript results if available
    try:
        with open('output/data_comparison_js.json', 'r') as f:
            js_output = json.load(f)

        print('\n' + '=' * 80)
        print('Comparison with JavaScript Results')
        print('=' * 80)

        # Compare overall statistics
        print('\nOverall Statistics Comparison:')
        print(f"  Total Values:  JS={js_output['statistics']['total_values']:,}, "
              f"PY={output['statistics']['total_values']:,}, "
              f"Match={'YES' if js_output['statistics']['total_values'] == output['statistics']['total_values'] else 'NO'}")
        print(f"  Valid Values:  JS={js_output['statistics']['valid_values']:,}, "
              f"PY={output['statistics']['valid_values']:,}, "
              f"Match={'YES' if js_output['statistics']['valid_values'] == output['statistics']['valid_values'] else 'NO'}")
        print(f"  Null Values:   JS={js_output['statistics']['null_values']:,}, "
              f"PY={output['statistics']['null_values']:,}, "
              f"Match={'YES' if js_output['statistics']['null_values'] == output['statistics']['null_values'] else 'NO'}")

        if js_output['statistics']['min'] is not None and output['statistics']['min'] is not None:
            min_diff = abs(js_output['statistics']['min'] - output['statistics']['min'])
            print(f"  Min:           JS={js_output['statistics']['min']:.4f}, "
                  f"PY={output['statistics']['min']:.4f}, "
                  f"Diff={min_diff:.6f}")

        if js_output['statistics']['max'] is not None and output['statistics']['max'] is not None:
            max_diff = abs(js_output['statistics']['max'] - output['statistics']['max'])
            print(f"  Max:           JS={js_output['statistics']['max']:.4f}, "
                  f"PY={output['statistics']['max']:.4f}, "
                  f"Diff={max_diff:.6f}")

        if js_output['statistics']['mean'] is not None and output['statistics']['mean'] is not None:
            mean_diff = abs(js_output['statistics']['mean'] - output['statistics']['mean'])
            print(f"  Mean:          JS={js_output['statistics']['mean']:.4f}, "
                  f"PY={output['statistics']['mean']:.4f}, "
                  f"Diff={mean_diff:.6f}")

        # Compare first radial value by value
        print('\nFirst Radial Comparison (first 10 values):')
        js_first = js_output['first_radial_full'][:10]
        py_first = output['first_radial_full'][:10]
        all_match = True

        for i, (js_val, py_val) in enumerate(zip(js_first, py_first)):
            if js_val is None and py_val is None:
                match = 'OK'
            elif js_val is None or py_val is None:
                match = 'DIFF'
                all_match = False
            else:
                diff = abs(js_val - py_val)
                if diff < 0.01:  # Allow small floating point differences
                    match = 'OK'
                else:
                    match = f'DIFF({diff:.4f})'
                    all_match = False

            js_str = 'null' if js_val is None else f"{js_val:.4f}"
            py_str = 'null' if py_val is None else f"{py_val:.4f}"
            print(f"  Gate {i:3d}: JS={js_str:10s}, PY={py_str:10s} [{match}]")

        print(f"\nFirst Radial Match: {'YES' if all_match else 'NO'}")

    except FileNotFoundError:
        print('\nJavaScript results not found. Run compare_data_values.js first.')

if __name__ == '__main__':
    compare_data_values()

File: test-verification/compare_data_values.js
/**
 * Deep Data Comparison - Compare actual REF values between JS and Python
 * Verifies that decompression and scaling produce identical numerical results
 */

import { read_nexrad_archive } from '../src/nexrad_archive.js';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Test with KHGX file (most comprehensive)
const TEST_FILE = 's3://unidata-nexrad-level2/2022/03/22/KHGX/KHGX20220322_120125_V06';

async function compareDataValues() {
    console.log('='.repeat(80));
    console.log('Deep Data Value Comparison: JavaScript vs Python');
    console.log('='.repeat(80));
    console.log(`\nTest File: ${TEST_FILE}\n`);

    // Read with JavaScript
    console.log('Reading with JavaScript...');
    const radar = await read_nexrad_archive(TEST_FILE);

    // Get REF data
    console.log('Extracting REF moment data...');
    const range_info = radar.get_range(0, 'REF');
    const max_ngates = range_info.length;
    const refData = radar.get_data('REF', max_ngates);

    console.log(`\nJavaScript REF Data Shape: [${refData.length}, ${max_ngates}]`);

    // Calculate statistics for all radials
    const stats = {
        total_radials: refData.length,
        total_gates: max_ngates,
        total_values: 0,
        valid_values: 0,
        null_values: 0,

        // Per-radial statistics
        radial_samples: [],

        // Overall statistics (excluding nulls)
        all_valid_values: [],
        min: null,
        max: null,
        mean: null,

        // Sample radials for detailed comparison
        sample_radials: [0, 100, 500, 1000, 5000, 10000]
    };

    // Collect statistics
    for (let radial = 0; radial < refData.length; radial++) {
        const radialData = refData[radial];
        let validCount = 0;
        let nullCount = 0;
        let sum = 0;
        let radialMin = null;
        let radialMax = null;

        for (let gate = 0; gate < radialData.length; gate++) {
            stats.total_values++;
            const value = radialData[gate];

            if (value === null || value === undefined || isNaN(value)) {
                nullCount++;
                stats.null_values++;
            } else {
                validCount++;
                stats.valid_values++;
                sum += value;
                stats.all_valid_values.push(value);

                if (radialMin === null || value < radialMin) radialMin = value;
                if (radialMax === null || value > radialMax) radialMax = value;
            }
        }

        // Store sample radials
        if (stats.sample_radials.includes(radial)) {
            stats.radial_samples.push({
                radial_index: radial,
                total_gates: radialData.length,
                valid_count: validCount,
                null_count: nullCount,
                min: radialMin,
                max: radialMax,
                mean: validCount > 0 ? sum / validCount : null,
                // First 20 values for detailed comparison
                values_sample: radialData.slice(0, 20),
                // Values at specific gates
                gate_samples: {
                    gate_0: radialData[0],
                    gate_100: radialData[100],
                    gate_500: radialData[500],
                    gate_1000: radialData[1000],
                    gate_1500: radialData[1500]
                }
            });
        }
    }

    // Calculate overall statistics (avoid spread operator for large arrays)
    if (stats.all_valid_values.length > 0) {
        stats.min = stats.all_valid_values[0];
        stats.max = stats.all_valid_values[0];
        let sum = 0;

        for (const val of stats.all_valid_values) {
            if (val < stats.min) stats.min = val;
            if (val > stats.max) stats.max = val;
            sum += val;
        }

        stats.mean = sum / stats.all_valid_values.length;

        // Clear the large array to free memory
        stats.all_valid_values = null;
    }

    // Print summary statistics
    console.log('\n' + '='.repeat(80));
    console.log('JavaScript REF Data Statistics');
    console.log('='.repeat(80));
    console.log(`Total Values: ${stats.total_values.toLocaleString()}`);
    console.log(`Valid Values: ${stats.valid_values.toLocaleString()} (${(stats.valid_values / stats.total_values * 100).toFixed(2)}%)`);
    console.log(`Null Values: ${stats.null_values.toLocaleString()} (${(stats.null_values / stats.total_values * 100).toFixed(2)}%)`);
    console.log(`\nOverall Statistics (valid values only):`);
    console.log(`  Min: ${stats.min?.toFixed(4)}`);
    console.log(`  Max: ${stats.max?.toFixed(4)}`);
    console.log(`  Mean: ${stats.mean?.toFixed(4)}`);

    // Print sample radials
    console.log('\n' + '='.repeat(80));
    console.log('Sample Radials (for Python comparison)');
    console.log('='.repeat(80));

    for (const sample of stats.radial_samples) {
        console.log(`\nRadial ${sample.radial_index}:`);
        console.log(`  Valid: ${sample.valid_count}, Null: ${sample.null_count}`);
        console.log(`  Min: ${sample.min?.toFixed(4)}, Max: ${sample.max?.toFixed(4)}, Mean: ${sample.mean?.toFixed(4)}`);
        console.log(`  First 10 values: [${sample.values_sample.slice(0, 10).map(v => v === null ? 'null' : v.toFixed(2)).join(', ')}]`);
        console.log(`  Gate samples:`);
        for (const [gate, value] of Object.entries(sample.gate_samples)) {
            console.log(`    ${gate}: ${value === null ? 'null' : value.toFixed(4)}`);
        }
    }

    // Save detailed results for Python comparison
    const output = {
        test_file: TEST_FILE,
        data_shape: [refData.length, max_ngates],
        statistics: {
            total_values: stats.total_values,
            valid_values: stats.valid_values,
            null_values: stats.null_values,
            min: stats.min,
            max: stats.max,
            mean: stats.mean
        },
        radial_samples: stats.radial_samples,
        // Include full first radial for exact comparison
        first_radial_full: refData[0]
    };

    const outputPath = path.join(__dirname, 'output', 'data_comparison_js.json');
    await fs.writeFile(outputPath, JSON.stringify(output, null, 2));
    console.log(`\n✓ Detailed results saved to: ${outputPath}`);

    // Print instructions for Python comparison
    console.log('\n' + '='.repeat(80));
    console.log('Next Step: Run Python comparison');
    console.log('='.repeat(80));
    console.log('Run: python compare_data_values.py');
    console.log('Then compare output/data_comparison_js.json with output/data_comparison_py.json');
}

compareDataValues().catch(err => {
    console.error('Error:', err);
    process.exit(1);
});
